### ISSUE TITLE
How to access JsonObject?

### ISSUE THREAD

**Issue by shericn** [kevin-montrose/Jil#109]
Hello,

I am trying the code below:

     dynamic result = JSON.DeserializeDynamic("{ \"key1\" : 3, \"key2\" : { \"nested\" : 4 } }");
     var key1 = result.key1;   // {3}
     var key2 = result.key2.nested;   // {4}

These are all expected. However, is there anyway to get the type of key1 and key2 after deserializing the JSON string? When I view the type of "result", I see "Jil.DeserializeDynamic.JsonObject". Looking through reflection, it seems like if JsonObject is public then it can expose lots of information to determine the type of "key1" and "key2". Are there anyway to deserialize an arbitrary JSON string using Jil and see if the children are primitives, array or object?

Another question: How's the performance of Jil when deserializing using DeserializeDynamic() function, as opposed to using other Json serializers such as Newtonsoft?

Thanks!

**Comment by kevin-montrose** [comment]
`JsonObject` is an implementation detail, it's not public because it has a habit of changing.  It's also not very pretty, frankly.

The classic way of determining if a `dynamic` is a particular type is to do trial casts, Jil itself has to do this when [serializing `dynamic`](https://github.com/kevin-montrose/Jil/blob/master/Jil/SerializeDynamic/DynamicSerializer.cs#L616) if it can't find a faster known interface.

However, thinking on it now... I'll investigate adding [`TypeConverter`](https://msdn.microsoft.com/en-us/library/system.componentmodel.typeconverter%28v=vs.110%29.aspx) support to the `dynamic`s returned from Jil.  That's fairly standard, wouldn't pollute interfaces, and keeps `JsonObject` tucked out of sight.

---

As for performance, I don't publish benchmark numbers for the `*Dynamic()` method because every library has a slightly different definition of dynamic (de)serialization.  Additionally, there are many valid "interpretations" of dynamic deserialization which makes an honest benchmark hard to make.

For example, Jil's `dynamic` does many conversions (such as to `DateTime` or `Guid`) on demand and doesn't cache the results.  This makes parsing faster, but access slower.  If the benchmark only deserialized, but didn't touch all properties, it would be biased in Jil's favor; on the other hand, if a property was touched many times it'd be biased against Jil.  Another example, Jil actually *does* read and process the whole stream before returning it's `dynamic`, but it's equally valid to do no parsing until an access happens; again you can construct benchmarks that favor either implementation.

Long story short, I think there's a fair chance any benchmark I throw together would be misleading.  Better for consumers to test their particular use case and decide.

All that being said, if you find a reasonable use case where Jil is slower than another library while serializing or deserializing dynamic objects I'd love to hear about it.  Jil's purpose is to be fast.

**Comment by shericn** [comment]
I see, thank you very much for the response. If TypeConverter is added, please let me know! I think that would be very helpful for me.
